### PR TITLE
adding json convertible type and necessary primitive conversions

### PR DESCRIPTION
**Pull request by LoganWright**
- [x] Add json convertible type
- [x] Update readme
- [x] Discuss implications of inclusion and possibly create separate package for json convertible inclusion as opt-in, also allows users to specify a Json library if desired

> Result of discussion.  There's no harm on this inclusion since it's completely disconnected from core.  If a user prefers something else, there's nothing preventing them. Inclusion makes it easier for users unsure of their preference

Helps w/ #28 and #23

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -75,6 +75,43 @@ Route.get("version") { request in
...
+
+#### Requests
+
+To access Json from parameters sent up with a request, use `request.json`
+
+```swift
+Route.post("messages") { request in
+	guard
+		   let json = request.json,
+			 let userId = json["userId"]?.stringValue,
```
Comment (tannernelson) [inline README.md]: If the request has a content type of `application/json` I think Vapor should automatically parse it into the `request.data`. This is how Laravel works. 

If the request is `{"test": 1}`

You would just do `let value = request.data["test"] // Optional(1)`

File: README.md [hunk 2]
```diff
@@ -75,6 +75,43 @@ Route.get("version") { request in
 
 This responds to all requests to `http://example.com/version` with the JSON dictionary `{"version": "1.0"}` and `Content-Type: application/json`.
 
+#### Complex Responses
+
+Vapor will do its best to infer a Json response from the returned type, but for some more complex examples, particularly on Linux, there are known inference issues.  To circumvent this, you can use the included `Json` type and wrap your response, for example:
```
Comment (tannernelson) [inline README.md]: Good explanation and example.

File: Sources/Vapor+Json.swift [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+/// To allow users to opt into preferred case style
+public typealias JSON = Json
+
+// MARK: Response
+
+extension Json : ResponseConvertible {
+    public func response() -> Response {
+        let js = serialize(.PrettyPrint)
+        return Response(status: .OK, text: js)
```
Comment (tannernelson) [inline Sources/Vapor+Json.swift]: The `Response` needs to have content type `.Json`.

```swift
let data = [UInt8](js.utf8)
return Response(status: .OK, data: data, type: .Json)
```

Something like that could work.

File: README.md (hunk 1 above)
Comment (czechboy0) [inline README.md]: How can you know here that it's a JSON object and not an array? Don't we need some conditional casting above into a dictionary type?

Comment (LoganWright) [inline README.md]: @czechboy0 conditional castings are nested within the Json enum, so you can use like an array or a dictionary safely.  `[StringKey]` will check if the underlying representation is `[String : Json]` and if so, pull that value.  `[IntKey]` will check if the underlying representation is `[Json]` and pull a value from that index if possible

Comment (czechboy0) [inline README.md]: 👍 smart

**Comment by LoganWright** [comment]
@tannernelson I was able to address the response comment, let me know how you'd like to proceed with automatic data conversion and I'll adjust that component as well

**Comment by tannernelson** [comment]
Did you determine if it's possible to use `NSJSON` on Linux for the request data parsing?

**Comment by LoganWright** [comment]
@tannernelson Just tried with the latest snapshot on Ubuntu 14 and was unable to get it to not crash, perhaps others have had luck.

**Comment by tannernelson** [comment]
@LoganWright Can you move this branch from your fork to a branch on qutheory/vapor? You should have write access now.

**Comment by LoganWright** [comment]
Moving to #46

**Pull request closed without merging**